x11/xauth.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package x11
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+)
+
+// See https://cgit.freedesktop.org/xorg/lib/libXau/tree/AuRead.c and
+// https://cgit.freedesktop.org/xorg/lib/libXau/tree/include/X11/Xauth.h
+// for details about the actual file format.
+type xauth struct {
+	Family  uint16
+	Address []byte
+	Number  []byte
+	Name    []byte
+	Data    []byte
+}
+
+func readBytes(f *os.File, data []byte) error {
+	n, err := f.Read(data)
+	if err != nil {
+		return err
+	}
+
+	if n != len(data) {
+		return fmt.Errorf("Could not read enough bytes")
+	}
+
+	return nil
+}
+
+func readChunk(f *os.File) ([]byte, error) {
+	// A chunk consists of a length encoded by two bytes and
+	// additional data which is the real value of the item
+	// we reading here from the file.
+
+	b := [2]byte{}
+	if err := readBytes(f, b[:]); err != nil {
+		return nil, err
+	}
+
+	size := int(binary.BigEndian.Uint16(b[:]))
+	chunk := make([]byte, size)
+	if err := readBytes(f, chunk); err != nil {
+		return nil, err
+	}
+
+	return chunk, nil
+}
+
+func (xa *xauth) readFromFile(f *os.File) error {
+	b := [2]byte{}
+	if err := readBytes(f, b[:]); err != nil {
+		return err
+	}
+	// The family field consists of two bytes
+	xa.Family = binary.BigEndian.Uint16(b[:])
+
+	var err error
+
+	if xa.Address, err = readChunk(f); err != nil {
+		return err
+	}
+
+	if xa.Number, err = readChunk(f); err != nil {
+		return err
+	}
+
+	if xa.Name, err = readChunk(f); err != nil {
+		return err
+	}
+
+	if xa.Data, err = readChunk(f); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ValidateXauthority validates a given Xauthority file. The file is valid
+// if it can be parsed and contains at least one cookie.
+func ValidateXauthority(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	return ValidateXauthorityFromFile(f)
+}
+
+// ValidateXauthority validates a given Xauthority file. The file is valid
+// if it can be parsed and contains at least one cookie.
+func ValidateXauthorityFromFile(f *os.File) error {
```

Comment:
Do we need to take a file, or is an io.Reader enough?

If so, I suggest naming this one as `ValidateXauthority` taking an `io.Reader`, and the other one as `ValidateXauthorityFile` taking a path.